Make proofing job asynchronous

zachmargolis opened the pull request:
**Why**:
Gives us more flexibility in case vendor requests take a long
period of time.

--

**Note**:
- The feature specs are failing, I need to make sure the `<meta>` tag for refreshing works in tests and the correct jobs run, am investigating now, but I'd love a review on the rest of the code

This is what the page looks like now 

<img width="711" alt="login_gov_" src="https://user-images.githubusercontent.com/458784/27971293-1f7b1a3e-6320-11e7-9ebb-a29096e06371.png">

monfresh commented on app/services/idv/vendor_result.rb:
```diff
@@ -14,8 +14,8 @@ def self.new_from_json(json)
       @success = success
-      @errors = errors
-      @reasons = reasons
+      @errors = errors || {}
```
If this need to be an empty hash by default, why not do it in the keyword arguments? Same with `reasons` below.

zachmargolis replied:
Good point, I'll update

monfresh commented on app/controllers/verify/finance_controller.rb:
```diff
@@ -7,13 +7,18 @@ class FinanceController < ApplicationController
     before_action :confirm_step_allowed
     before_action :submit_idv_form, only: [:create]
     before_action :submit_idv_job, only: [:create]
+    before_action :refresh_if_not_ready, only: [:show]
```
Why not attach this to the existing "new" action? That way `create` can stay as it was and you wouldn't need to `redirect_to verify_finance_result_path`. Same with the other controllers.

monfresh replied:
I think I misunderstood. I thought each page was checking for the result before the page loaded, but that check is happening after each step is submitted. I was confused by the `show` action because it implies a GET request, but it's submitting a form. Can you think of a better way to handle this?

monfresh replied:
How about a new controller for each step that polls for the vendor result, and then proceeds with processing the result? So, after the SessionsController submits the data to the vendor, it would redirect to this new Job Polling controller, and when the result is ready, it will execute what we used to have in the `create` action, such as:
```ruby
result = step.submit
analytics.track_event(Analytics::IDV_BASIC_INFO_SUBMITTED_VENDOR, result.to_h)

if result.success?
  process_success
else
  process_failure
end
```
That way, we clean up the controller by removing `before_action`s. I'm thinking SessionsController#create will now look something like this:
```ruby
def create
  initialize_idv_session

  result = idv_form.submit(profile_params)
  analytics.track_event(Analytics::IDV_BASIC_INFO_SUBMITTED_FORM, result.to_h)

  if result.success?
    submit_idv_job
    redirect_to verify_session_job_polling_path
  else
    process_failure
  end
end
```
which is a lot more readable to me than having to look at `before_action`s to figure out what is going on.

Thoughts?

zachmargolis replied:
I find more controllers harder to follow, especially when the results are so tightly connected between actions like this.

I do agree about having fewer `before_actions` -- that was mostly in line with keeping the diff small.

I'll see what it looks like to extract a separate controller for this

zachmargolis replied:
Ok so I tried extracting into a controller, and due to the mixins we use and the amount of shared methods between the different views (ex `step` and `idv_form`), it did not separate cleanly. I was able to remove most of the `before_action`s by integrating them, so I'm pushing that change up now: f627bc703f8fce506dbd7943b51de09e234d002b

zachmargolis commented:
@monfresh @jmhooper ok I've fixed the last 2 failing acceptance tests, and then squashed everything up. I also added handling for when we don't hear back from the background job after a certain period of time (since if the jobs silently failed we would just spin forever otherwise)

It ends up using a generic error message like this, but we have tracking that it's due to a timeout

<img width="559" alt="screen shot 2017-07-18 at 10 57 52 am" src="https://user-images.githubusercontent.com/458784/28324587-ca6c4dd2-6ba9-11e7-9827-c92414850e3c.png">

monfresh commented on app/controllers/concerns/idv_session.rb:
```diff
@@ -42,6 +42,29 @@ def idv_attempter
+
+  def vendor_result_timed_out?
+    started_at = idv_session.async_result_started_at
+    return if started_at.blank?
```
Should we return `false` since this is a predicate method?

zachmargolis replied:
yes, that's a good point

monfresh commented on app/controllers/verify/sessions_controller.rb:
```diff
@@ -20,6 +19,18 @@ def new
+      end
+    end
+
+    def show
```
The name `show` is still throwing me off. When I see `show`, I think of a page that is displaying something to the user, and not doing anything else. Maybe rename this to `result`?

I also have a question about what happens when the vendor times out. Let's say the user submits the initial personal info form, which redirects to `/verify/session/result` while we're waiting for the vendor to respond. If the vendor times out, we are currently failing the user, which is not fair to the user. If the user clicks "Try again" on the modal, they are sent back to the personal info page, but the URL hasn't changed. It's still `/verify/session/result`. So, if the user refreshes the page, it will trigger another attempt, and they will now have 1 less attempt remaining. Can we change this behavior such that the user is not penalized, and is instead presented with a message stating what happened and to try again later?

monfresh replied:
or `process_vendor_result`?

zachmargolis replied:
I thought we were sticking to the Rails convention of using only built-in action names as much as possible (`index`, `edit`, `show`, `create`, etc etc). To me, this is showing the result of the vendor result, which sometimes is in progress (hence rendering a spinner).

Re: when the vendor times out, I figured that incrementing the count was acceptable for a few reasons:
1. if a vendor is permanently down, we should not continue to spam them with requests and waste their time
2. if the request did get to the vendor, then it counts towards our limit on their side anyways

I don't think the experience of timing out is ideal, but I think that treating timeout errors as negative responses is acceptable.

Re messaging: I'll look into adding a "our request to verify your information timed out" error message.

zachmargolis replied:
Added error messaging in 516c32f8a9dfdcbac8c6a3db86b148daece27235

monfresh replied:
I'll have to think about this some more, but one thing we can say for sure is that this `show` method is different from all the other `show` methods we currently have in that it is acting a lot like a `create` method since it's processing a form and acting on that submission as opposed to just rendering a view.

As for the vendor limits, my understanding was that the limits were to prevent too many failed attempts by a user. If the vendor is only temporarily down for 5 minutes, users will be prevented from trying again for another 24 hours if we count the attempts against them. Perhaps we should open up this discussion to the rest of the team? I'm not sure what would be the ideal way to handle this.

zachmargolis replied:
Yeah I think we can talk about it with the rest of the team, and since LOA3 has not launched publicly yet, would you feel OK if we merged it with the logic as-is and followed up with how to handle these other edge cases?

monfresh commented on app/controllers/concerns/idv_session.rb:
```diff
@@ -42,6 +42,30 @@ def idv_attempter
+  def timed_out_vendor_error
+    Idv::VendorResult.new(
+      success: false,
+      errors: { timed_out: ['Timed out waiting for vendor response'] },
```
Should we localize this string?

zachmargolis replied:
I thought about that, but then I realized that it only is sent to error analytics for us to consume internally -- it's not user-facing at all, so I think it's OK to leave un-translated

monfresh replied:
I was thinking about someone in France, for example, forking our code to deploy over there. They might want to have analytics in French? Then again, we don't translate any of the analytics event names, so maybe skip this for now?

zachmargolis replied:
Yeah all of our analytics is in English, I agree we should skip for now.

monfresh commented on spec/controllers/verify/finance_controller_spec.rb:
```diff
@@ -80,32 +80,182 @@
+            success: false,
+            errors: { ccn: ['The ccn could not be verified.'] }
+          )
+        end
```
This `result` variable doesn't seem to be used in this context.

zachmargolis replied:
This is used in the `before` block above to be stored in redis

```
        VendorValidatorResultStorage.new.store(result_id: result_id, result: result)
```

zachmargolis replied:
The old tests relied on being able to give bad values directly to the vendorvalidator, but now that we have split this into a separate step, we need to save the results in redis for each step

monfresh replied:
I see it now. It's used above in the `before` block, which was confusing. I generally expect `let` to come first, and then have it used below its definition.

zachmargolis replied:
The `let` does come first, but we're overriding it for this one case :)
